{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c889a4c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome, ram.ayyala!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# CELL 1: LIBRARIES & IMPORTS\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json\n",
    "import re\n",
    "import os\n",
    "import yaml\n",
    "import glob\n",
    "from pathlib import Path\n",
    "from datetime import datetime\n",
    "import synapseclient\n",
    "from synapseclient.models import (\n",
    "    Column, ColumnType, Dataset, EntityRef, File, Folder, Project, FacetType, DatasetCollection, JSONSchema, SchemaOrganization\n",
    ")\n",
    "import time\n",
    "from pprint import pprint\n",
    "from synapseclient.core.utils import make_bogus_data_file\n",
    "\n",
    "# 1. Set up Synapse Python client\n",
    "syn = synapseclient.Synapse()\n",
    "syn.login()\n",
    "from typing import Dict, List, Any, Set, Union\n",
    "from synapseclient import Wiki\n",
    "from collections import defaultdict, Counter\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1c33b50d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace your own project name here\n",
    "PROJECT_ENT = Project(name=\"Mock ALS Project\").get()\n",
    "# Replace your own json schema organization name here\n",
    "ORG_NAME = \"ampals.schemas\"\n",
    "VERSION = \"0.0.1\"\n",
    "NEW_VERSION = \"0.0.2\"\n",
    "\n",
    "SCHEMA_NAME = \"OmicFile\"\n",
    "\n",
    "title = \"ALS Omic File Schema\"\n",
    "schema_body = {\n",
    "  \"$schema\": \"https://json-schema.org/draft-07/schema\",\n",
    "  \"$id\": \"https://repo-prod.prod.sagebase.org/repo/v1/schema/type/registered/org.synapse.ampals-omic-file\",\n",
    "  \"title\": \"OmicFile\",\n",
    "  \"type\": \"object\",\n",
    "  \"description\": \"File containing omic data such as sequencing, expression, or proteomics data\",\n",
    "  \"additionalProperties\": {},\n",
    "  \"properties\": {\n",
    "    \"FACSPopulation\": {\n",
    "      \"description\": \"FACS-sorted cell population description\",\n",
    "      \"title\": \"FACS Population\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"GEOSuperSeries\": {\n",
    "      \"description\": \"GEO SuperSeries identifier\",\n",
    "      \"title\": \"GEO SuperSeries\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"alternateName\": {\n",
    "      \"description\": \"An alternate name that can be used for search and discovery improvement.\",\n",
    "      \"title\": \"Alternate Name\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"analysisTypes\": {\n",
    "      \"description\": \"Computational and statistical analysis methods applied to the data\",\n",
    "      \"items\": {\n",
    "        \"description\": \"Computational and statistical analysis methods applied to the data\",\n",
    "        \"enum\": [\n",
    "          \"Normalized Protein Expression (NPX)\",\n",
    "          \"GeneSetEnrichmentAnalysis\",\n",
    "          \"differential_expression\",\n",
    "          \"pathway_enrichment\",\n",
    "          \"network_analysis\",\n",
    "          \"proteomics_analysis\",\n",
    "          \"metabolomics_analysis\",\n",
    "          \"longitudinal_analysis\",\n",
    "          \"survival_analysis\",\n",
    "          \"meta_analysis\",\n",
    "          \"correlation_analysis\",\n",
    "          \"multivariate_analysis\",\n",
    "          \"machine_learning\"\n",
    "        ],\n",
    "        \"title\": \"AnalysisMethodEnum\",\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Analysis Types\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"annotations\": {\n",
    "      \"description\": \"Types of biological databases and annotations referenced in this file\",\n",
    "      \"items\": {\n",
    "        \"description\": \"Types of biological databases and annotation references\",\n",
    "        \"enum\": [\n",
    "          \"KEGG\",\n",
    "          \"GO\",\n",
    "          \"UniProt\",\n",
    "          \"Ensembl\",\n",
    "          \"HGNC\",\n",
    "          \"RefSeq\",\n",
    "          \"MONDO\",\n",
    "          \"HPO\",\n",
    "          \"SNOMED\",\n",
    "          \"ICD\",\n",
    "          \"pathway_enrichment\",\n",
    "          \"protein_pathway\",\n",
    "          \"metabolic_pathway\",\n",
    "          \"signaling_pathway\"\n",
    "        ],\n",
    "        \"title\": \"AnnotationTypeEnum\",\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Biological Annotations\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"assay\": {\n",
    "      \"description\": \"What's used to generate data in file, typically referring to the assay. Omit if not applicable.\",\n",
    "      \"items\": {\n",
    "        \"description\": \"\",\n",
    "        \"enum\": [\n",
    "          \"2D AlamarBlue absorbance\",\n",
    "          \"2D AlamarBlue fluorescence\",\n",
    "          \"3D confocal imaging\",\n",
    "          \"3D electron microscopy\",\n",
    "          \"3D imaging\",\n",
    "          \"3D microtissue viability\",\n",
    "          \"actigraphy\",\n",
    "          \"AlgometRx Nociometer\",\n",
    "          \"auditory brainstem response\",\n",
    "          \"ATAC-seq\",\n",
    "          \"ATPase activity assay\",\n",
    "          \"BrdU proliferation assay\",\n",
    "          \"Bru-seq\",\n",
    "          \"BruChase-seq\",\n",
    "          \"CAPP-seq\",\n",
    "          \"CUT&RUN\",\n",
    "          \"ChIP-seq\",\n",
    "          \"Child Behavior Checklist for Ages 1.5-5\",\n",
    "          \"Child Behavior Checklist for Ages 6-18\",\n",
    "          \"CODEX\",\n",
    "          \"CROP-seq\",\n",
    "          \"confocal microscopy\",\n",
    "          \"Corsi blocks\",\n",
    "          \"current clamp assay\",\n",
    "          \"DART-seq\",\n",
    "          \"diffusion MRI\",\n",
    "          \"distortion product otoacoustic emissions\",\n",
    "          \"DNA optical mapping\",\n",
    "          \"ELISA\",\n",
    "          \"ERR bisulfite sequencing\",\n",
    "          \"EdU proliferation assay\",\n",
    "          \"FIA-MSMS\",\n",
    "          \"FLIPR high-throughput cellular screening\",\n",
    "          \"Fluorescence In Situ Hybridization\",\n",
    "          \"Focus group\",\n",
    "          \"FTIR spectroscopy\",\n",
    "          \"HI-C\",\n",
    "          \"HPLC\",\n",
    "          \"Interview\",\n",
    "          \"ISO-seq\",\n",
    "          \"MIB/MS\",\n",
    "          \"MeRIP-seq\",\n",
    "          \"Matrigel-based tumorigenesis assay\",\n",
    "          \"MudPIT\",\n",
    "          \"NIH Toolbox\",\n",
    "          \"NOMe-seq\",\n",
    "          \"RNA array\",\n",
    "          \"RNA-seq\",\n",
    "          \"RPPA\",\n",
    "          \"Riccardi and Ablon scales\",\n",
    "          \"SNP array\",\n",
    "          \"SUSHI\",\n",
    "          \"Sanger sequencing\",\n",
    "          \"Social Responsiveness Scale\",\n",
    "          \"Social Responsiveness Scale, Second Edition\",\n",
    "          \"SomaScan\",\n",
    "          \"T cell receptor repertoire sequencing\",\n",
    "          \"TIDE\",\n",
    "          \"TMT quantitation\",\n",
    "          \"TriKinetics activity monitoring\",\n",
    "          \"Von Frey test\",\n",
    "          \"active avoidance learning behavior assay\",\n",
    "          \"array\",\n",
    "          \"atomic force microscopy\",\n",
    "          \"autoradiography\",\n",
    "          \"bisulfite sequencing\",\n",
    "          \"blood chemistry measurement\",\n",
    "          \"blue native PAGE\",\n",
    "          \"body size trait measurement\",\n",
    "          \"bone histomorphometry\",\n",
    "          \"brightfield microscopy\",\n",
    "          \"cAMP-Glo Max Assay\",\n",
    "          \"CAGE-seq\",\n",
    "          \"calcium retention capacity assay\",\n",
    "          \"cell competition\",\n",
    "          \"cell count\",\n",
    "          \"cell painting\",\n",
    "          \"cell proliferation\",\n",
    "          \"cell viability assay\",\n",
    "          \"clinical data\",\n",
    "          \"CLIP-seq\",\n",
    "          \"cNF-Skindex\",\n",
    "          \"cognitive assessment\",\n",
    "          \"combination library screen\",\n",
    "          \"combination screen\",\n",
    "          \"complex II enzyme activity assay\",\n",
    "          \"compound screen\",\n",
    "          \"contextual conditioning behavior assay\",\n",
    "          \"conventional MRI\",\n",
    "          \"Children's Dermatology Life Quality Index Questionnaire\",\n",
    "          \"differential scanning calorimetry\",\n",
    "          \"dynamic light scattering\",\n",
    "          \"electrochemiluminescence\",\n",
    "          \"electrophoretic light scattering\",\n",
    "          \"elevated plus maze test\",\n",
    "          \"FACE-Q Appearance-related Distress\",\n",
    "          \"FACS\",\n",
    "          \"flow cytometry\",\n",
    "          \"focus forming assay\",\n",
    "          \"functional MRI\",\n",
    "          \"gait measurement\",\n",
    "          \"gel filtration chromatography\",\n",
    "          \"gel permeation chromatography\",\n",
    "          \"genotyping\",\n",
    "          \"high content screen\",\n",
    "          \"high frequency ultrasound\",\n",
    "          \"high-performance liquid chromatography/tandem mass spectrometry\",\n",
    "          \"immunoassay\",\n",
    "          \"immunocytochemistry\",\n",
    "          \"immunofluorescence\",\n",
    "          \"immunohistochemistry\",\n",
    "          \"in silico synthesis\",\n",
    "          \"in vitro tumorigenesis\",\n",
    "          \"in vivo PDX viability\",\n",
    "          \"in vivo bioluminescence\",\n",
    "          \"in vivo tumor growth\",\n",
    "          \"jumping library\",\n",
    "          \"LCM\",\n",
    "          \"label free mass spectrometry\",\n",
    "          \"laser speckle imaging\",\n",
    "          \"light scattering assay\",\n",
    "          \"liquid chromatography-electrochemical detection\",\n",
    "          \"liquid chromatography/mass spectrometry\",\n",
    "          \"liquid chromatography/tandem mass spectrometry\",\n",
    "          \"lncRNA-seq\",\n",
    "          \"local field potential recording\",\n",
    "          \"long term potentiation assay\",\n",
    "          \"mRNA counts\",\n",
    "          \"m6A array\",\n",
    "          \"magnetic resonance angiography\",\n",
    "          \"magnetic resonance spectroscopy\",\n",
    "          \"Magnetization-Prepared Rapid Gradient Echo MRI\",\n",
    "          \"mass spectrometry\",\n",
    "          \"massively parallel reporter assay\",\n",
    "          \"metabolic screening\",\n",
    "          \"methylation array\",\n",
    "          \"miRNA array\",\n",
    "          \"miRNA-seq\",\n",
    "          \"microrheology\",\n",
    "          \"Skindex-16\",\n",
    "          \"multi-electrode array\",\n",
    "          \"nanoparticle tracking analysis\",\n",
    "          \"NanoString nCounter Analysis System\",\n",
    "          \"n-back task\",\n",
    "          \"neuropsychological assessment\",\n",
    "          \"next generation targeted sequencing\",\n",
    "          \"novelty response behavior assay\",\n",
    "          \"open field test\",\n",
    "          \"optical tomography\",\n",
    "          \"optical coherence tomography\",\n",
    "          \"optokinetic reflex assay\",\n",
    "          \"oscillatory rheology\",\n",
    "          \"oxBS-seq\",\n",
    "          \"oxygen consumption assay\",\n",
    "          \"pattern electroretinogram\",\n",
    "          \"perineurial cell thickness\",\n",
    "          \"pharmocokinetic ADME assay\",\n",
    "          \"phase-contrast microscopy\",\n",
    "          \"photograph\",\n",
    "          \"polymerase chain reaction\",\n",
    "          \"polysomnography\",\n",
    "          \"positron emission tomography\",\n",
    "          \"PROMIS Cognitive Function\",\n",
    "          \"Olink Proximity Extension Assay\",\n",
    "          \"pure tone average\",\n",
    "          \"quantitative PCR\",\n",
    "          \"Quant-seq\",\n",
    "          \"questionnaire\",\n",
    "          \"reactive oxygen species assay\",\n",
    "          \"reporter gene assay\",\n",
    "          \"rheometry\",\n",
    "          \"RIP-seq\",\n",
    "          \"ribo-seq\",\n",
    "          \"rotarod performance test\",\n",
    "          \"sandwich ELISA\",\n",
    "          \"scCGI-seq\",\n",
    "          \"scale\",\n",
    "          \"SaferSeqS\",\n",
    "          \"single molecule drug screen assay\",\n",
    "          \"single-cell RNA-seq\",\n",
    "          \"single cell ATAC-seq\",\n",
    "          \"snATAC-seq\",\n",
    "          \"snRNA-seq\",\n",
    "          \"single-nucleus RNA-seq\",\n",
    "          \"small molecule library screen\",\n",
    "          \"sorbitol dehydrogenase activity level assay\",\n",
    "          \"spatial frequency domain imaging\",\n",
    "          \"spatial transcriptomics\",\n",
    "          \"static histomorphometry\",\n",
    "          \"static light scattering\",\n",
    "          \"survival\",\n",
    "          \"targeted exome sequencing\",\n",
    "          \"traction force microscopy\",\n",
    "          \"transcranial doppler ultrasonography\",\n",
    "          \"twin spot assay\",\n",
    "          \"ultra high-performance liquid chromatography/tandem mass spectrometry\",\n",
    "          \"western blot\",\n",
    "          \"whole exome sequencing\",\n",
    "          \"whole genome sequencing\",\n",
    "          \"whole-cell patch clamp\",\n",
    "          \"word recognition score\",\n",
    "          \"STR profile\",\n",
    "          \"Extension control 1\",\n",
    "          \"Incubation control 1\",\n",
    "          \"Amplification control 1\",\n",
    "          \"Protein target assay\"\n",
    "        ],\n",
    "        \"title\": \"Assay\",\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Measurement Technique\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"biospecimenType\": {\n",
    "      \"description\": \"Type of biological specimen(s) in the file.\",\n",
    "      \"items\": {\n",
    "        \"description\": \"\",\n",
    "        \"enum\": [\n",
    "          \"blood\",\n",
    "          \"plasma\",\n",
    "          \"serum\",\n",
    "          \"csf\",\n",
    "          \"brain_tissue\",\n",
    "          \"spinal_cord\",\n",
    "          \"muscle\",\n",
    "          \"fibroblast\",\n",
    "          \"ipsc\",\n",
    "          \"motor_neuron\",\n",
    "          \"unknown\"\n",
    "        ],\n",
    "        \"title\": \"BiospecimenTypeEnum\",\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Biospecimen Type\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"cellType\": {\n",
    "      \"description\": \"Type of cell or cell line used in the file.\",\n",
    "      \"items\": {\n",
    "        \"description\": \"\",\n",
    "        \"enum\": [\n",
    "          \"motor_neuron\",\n",
    "          \"astrocyte\",\n",
    "          \"microglia\",\n",
    "          \"oligodendrocyte\",\n",
    "          \"fibroblast\",\n",
    "          \"pbmc\",\n",
    "          \"whole_blood\",\n",
    "          \"unknown\"\n",
    "        ],\n",
    "        \"title\": \"CellTypeEnum\",\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Cell Type\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"citation\": {\n",
    "      \"description\": \"identifies academic articles that are recommended by the data provider be cited in addition to the file itself.\",\n",
    "      \"items\": {\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Citation\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"collection\": {\n",
    "      \"description\": \"Associated collection(s)\",\n",
    "      \"items\": {\n",
    "        \"description\": \"Original collection, e.g. the consortium that the dataset comes from.\",\n",
    "        \"enum\": [\n",
    "          \"ALS Knowledge Portal\",\n",
    "          \"Answer ALS\",\n",
    "          \"CReATe Consortium\",\n",
    "          \"Genomic Translation for ALS Care (GTAC)\",\n",
    "          \"New York Genome Center ALS Consortium\",\n",
    "          \"National Institutes of Health\",\n",
    "          \"Project MinE\",\n",
    "          \"ALS Compute\",\n",
    "          \"amp_als\",\n",
    "          \"target_als\",\n",
    "          \"all_als\",\n",
    "          \"c_path\",\n",
    "          \"alsdb\",\n",
    "          \"geo\",\n",
    "          \"sra\",\n",
    "          \"als_compute\"\n",
    "        ],\n",
    "        \"title\": \"CollectionEnum\",\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Collection\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"contributor\": {\n",
    "      \"description\": \"Institution or person responsible for collecting and managing the files and records in the file.\",\n",
    "      \"items\": {\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Contributor(s)\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"creator\": {\n",
    "      \"description\": \"Properly formatted name of the organization or person that is creator of the file (e.g. \\\"Sage Bionetworks\\\" or \\\"John Doe\\\"), not an id. Most preferred is the project PI or data lead who's responsible for its conception and initial content creation.\",\n",
    "      \"items\": {\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Creator\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"dataType\": {\n",
    "      \"description\": \"Omic data types (genomics, transcriptomics, proteomics, etc.)\",\n",
    "      \"items\": {\n",
    "        \"description\": \"Types of omic data contained in files\",\n",
    "        \"enum\": [\n",
    "          \"raw_sequencing\",\n",
    "          \"aligned_reads\",\n",
    "          \"variant_calls\",\n",
    "          \"genome_annotation\",\n",
    "          \"gene_expression\",\n",
    "          \"transcript_counts\",\n",
    "          \"differential_expression\",\n",
    "          \"splice_variants\",\n",
    "          \"methylation\",\n",
    "          \"chromatin_accessibility\",\n",
    "          \"histone_modification\",\n",
    "          \"protein_abundance\",\n",
    "          \"peptide_identification\",\n",
    "          \"protein_modification\",\n",
    "          \"metabolite_abundance\",\n",
    "          \"metabolic_pathways\",\n",
    "          \"integrated_omics\",\n",
    "          \"network_analysis\",\n",
    "          \"raw_data\",\n",
    "          \"processed_data\",\n",
    "          \"analyzed_results\",\n",
    "          \"proteomics\",\n",
    "          \"geneExpression\",\n",
    "          \"transcriptomics\",\n",
    "          \"transcriptomics_nascent\",\n",
    "          \"epigenomics\",\n",
    "          \"metabolomics\",\n",
    "          \"kinomics\",\n",
    "          \"isoformExpression\",\n",
    "          \"chromatinActivity\",\n",
    "          \"AlignedReads\",\n",
    "          \"genomicVariants\",\n",
    "          \"genomicFeatures\",\n",
    "          \"SomaticVariants\",\n",
    "          \"GermlineVariants\",\n",
    "          \"CopyNumberVariants\",\n",
    "          \"StructuralVariants\",\n",
    "          \"AnnotatedSomaticVariants\",\n",
    "          \"AnnotatedGermlineVariants\",\n",
    "          \"raw_counts\",\n",
    "          \"RawIntensities\",\n",
    "          \"NormalizedIntensities\"\n",
    "        ],\n",
    "        \"title\": \"OmicDataTypeEnum\",\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Data Type\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"description\": {\n",
    "      \"description\": \"Blurb for the file; should be no more than 500 characters.\",\n",
    "      \"title\": \"Description\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"disease\": {\n",
    "      \"description\": \"Disease(s) or condition(s) studied in the file.\",\n",
    "      \"items\": {\n",
    "        \"description\": \"\",\n",
    "        \"enum\": [\n",
    "          \"ALS\",\n",
    "          \"FTD\",\n",
    "          \"Control\",\n",
    "          \"Unknown\"\n",
    "        ],\n",
    "        \"title\": \"DiseaseEnum\",\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Disease\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"fileFormat\": {\n",
    "      \"description\": \"Format of the omic data file (FASTQ, BAM, VCF, etc.)\",\n",
    "      \"title\": \"File Format\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"includedInDataCatalog\": {\n",
    "      \"description\": \"Link(s) to known data catalog(s) the file is included in.\",\n",
    "      \"title\": \"Included In Data Catalog\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"individualCount\": {\n",
    "      \"description\": \"Number of unique individuals included in the file (whether as individual-level or as aggregate data). Omit if not applicable/unknown.\",\n",
    "      \"title\": \"Individual Count\",\n",
    "      \"type\": \"integer\"\n",
    "    },\n",
    "    \"keywords\": {\n",
    "      \"items\": {\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Keywords\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"libraryLayout\": {\n",
    "      \"description\": \"Single-end or paired-end sequencing layout for files containing sequencing data.\",\n",
    "      \"items\": {\n",
    "        \"description\": \"\",\n",
    "        \"enum\": [\n",
    "          \"single\",\n",
    "          \"paired\"\n",
    "        ],\n",
    "        \"title\": \"LibraryLayoutEnum\",\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Library Layout\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"libraryStrategy\": {\n",
    "      \"description\": \"Sequencing library strategy (RNA-seq, WGS, ChIP-seq, etc.)\",\n",
    "      \"title\": \"Library Strategy\"\n",
    "    },\n",
    "    \"license\": {\n",
    "      \"description\": \"Unless information for license is clear, this should default to UNKNOWN.\",\n",
    "      \"title\": \"License\"\n",
    "    },\n",
    "    \"originalSampleName\": {\n",
    "      \"description\": \"Original sample name from data source\",\n",
    "      \"title\": \"Original Sample Name\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"platform\": {\n",
    "      \"description\": \"Sequencing or analysis platform used\",\n",
    "      \"title\": \"Platform\",\n",
    "      \"enum\": [\n",
    "        \"7T Bruker Biospec\",\n",
    "        \"10x Visium Spatial Gene Expression\",\n",
    "        \"2D CellTiter-Glo\",\n",
    "        \"2D Incucyte\",\n",
    "        \"Affymetrix Genome-Wide Human SNP 5.0 Array\",\n",
    "        \"Affymetrix Genome-Wide Human SNP 6.0 Array\",\n",
    "        \"Affymetrix Human Gene 1.0 ST Array\",\n",
    "        \"Affymetrix Human Genome U133 Plus 2.0 Array\",\n",
    "        \"Affymetrix U133AB\",\n",
    "        \"Agilent 44Karray\",\n",
    "        \"BD FACS Calibur\",\n",
    "        \"BD FACSymphony\",\n",
    "        \"BGISEQ-500\",\n",
    "        \"Bionano Irys\",\n",
    "        \"Caliper\",\n",
    "        \"Cherry Imaging FACE Platform\",\n",
    "        \"Cherry Imaging TRACE Platform\",\n",
    "        \"Chromium X\",\n",
    "        \"EnVision 2103 Multiplate Reader\",\n",
    "        \"GE Discovery MR750 3T\",\n",
    "        \"GE Optima MR450W 1.5T\",\n",
    "        \"GE Signa HDxt 1.5T\",\n",
    "        \"GE Signa Genesis 1.5T\",\n",
    "        \"GE Signa HDxt 3T\",\n",
    "        \"GE Signa Premier 3T\",\n",
    "        \"GE Signa Excite 1.5T\",\n",
    "        \"Hitachi Echelon 1.5T\",\n",
    "        \"Hitachi Oasis 1.2T\",\n",
    "        \"IVIS Spectrum In Vivo Imaging System\",\n",
    "        \"Illumina 1M\",\n",
    "        \"Illumina Genome Analyzer IIx\",\n",
    "        \"Illumina HiSeq 2000\",\n",
    "        \"Illumina HiSeq 2500\",\n",
    "        \"Illumina HiSeq 3000\",\n",
    "        \"Illumina HiSeq 4000\",\n",
    "        \"Illumina HiSeq X\",\n",
    "        \"Illumina Human660W-Quad v1.0 BeadChip\",\n",
    "        \"Illumina HumanHap300\",\n",
    "        \"Illumina HumanMethylation450\",\n",
    "        \"Illumina HumanOmni1-Quadv1.0\",\n",
    "        \"Illumina HumanOmniExpress-24 v1.0 BeadChip\",\n",
    "        \"Illumina HumanOmniExpress-24 v1.2 BeadChip\",\n",
    "        \"Illumina Infinium MethylationEPIC BeadChip v1.0 (850k)\",\n",
    "        \"Illumina Infinium MethylationEPIC BeadChip v2.0 (935k)\",\n",
    "        \"Illumina MiSeq\",\n",
    "        \"Illumina MouseWG-6 v2.0 expression beadchip\",\n",
    "        \"Illumina NextSeq 1000\",\n",
    "        \"Illumina NextSeq 2000\",\n",
    "        \"Illumina NextSeq 500\",\n",
    "        \"Illumina NextSeq 550\",\n",
    "        \"Illumina NovaSeq 6000\",\n",
    "        \"Illumina NovaSeq X\",\n",
    "        \"Illumina NovaSeq X Plus\",\n",
    "        \"Illumina Omni2pt5M\",\n",
    "        \"Illumina Omni5M\",\n",
    "        \"Illumina WholeGenome DASL\",\n",
    "        \"Illumina h650\",\n",
    "        \"Infinium HumanOmniExpressExome\",\n",
    "        \"LI-COR Odyssey CLx\",\n",
    "        \"LTQ Orbitrap XL\",\n",
    "        \"Leica Aperio AT2\",\n",
    "        \"Leica MZ16\",\n",
    "        \"Leica S9 Stereomicroscope\",\n",
    "        \"LifeViz Infinity System\",\n",
    "        \"LifeViz Micro System\",\n",
    "        \"MGI T-series\",\n",
    "        \"Malvern Zetasizer\",\n",
    "        \"NanoFCM\",\n",
    "        \"NanoString Human nCounter PanCancer IO360 Panel\",\n",
    "        \"Nanostring Counter\",\n",
    "        \"Nanostring GeoMx\",\n",
    "        \"Not Applicable\",\n",
    "        \"Olympus DP80\",\n",
    "        \"Olympus IX73\",\n",
    "        \"Orbitrap Fusion Lumos Tribrid\",\n",
    "        \"Other Platform\",\n",
    "        \"Oxford Nanopore\",\n",
    "        \"PacBio RS II\",\n",
    "        \"PacBio Sequel II System\",\n",
    "        \"PacBio Sequel IIe System\",\n",
    "        \"Pannoramic 250 Flash\",\n",
    "        \"Perlegen 300Karray\",\n",
    "        \"Philips Achieva 1.5T\",\n",
    "        \"Philips Achieva 3T\",\n",
    "        \"Philips FEI Tecnai 12\",\n",
    "        \"Philips Intera Achieva 3T\",\n",
    "        \"Philips Ingenia 1.5T\",\n",
    "        \"Philips Ingenia 3T\",\n",
    "        \"Philips Panorama 1.0T\",\n",
    "        \"Promega GloMax Discover\",\n",
    "        \"Q Exative HF\",\n",
    "        \"Scale\",\n",
    "        \"Siemens Avanto 1.5T\",\n",
    "        \"Siemens Avanto Fit 1.5T\",\n",
    "        \"Siemens Magnetom Aera 1.5T\",\n",
    "        \"Siemens Magnetom Espree 1.5T\",\n",
    "        \"Siemens Magnetom Prisma 3T\",\n",
    "        \"Siemens Magnetom Skyra 3T\",\n",
    "        \"Siemens Magnetom Trio 3T\",\n",
    "        \"Siemens Magnetom Verio 3T\",\n",
    "        \"Siemens Magnetom Prisma Fit 3T\",\n",
    "        \"Spectramax M Series\",\n",
    "        \"TOOsonix System ONE-M\",\n",
    "        \"Toshiba Vantage Titan 1.5T\",\n",
    "        \"Varioskan LUX\",\n",
    "        \"Vectra H1 3D Imaging System\",\n",
    "        \"Ventana Benchmark XT\",\n",
    "        \"Vevo 3100 Imaging System\",\n",
    "        \"XF24 Extracellular Flux Analyzer\",\n",
    "        \"Zeiss LSM\",\n",
    "        \"Zeiss LSM 700\",\n",
    "        \"Zeiss LSM 980\",\n",
    "        \"Zeno Electronic Walkway\",\n",
    "        \"ZetaView\"\n",
    "      ],\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"processingLevel\": {\n",
    "      \"description\": \"Level of data processing (raw, processed, analyzed)\",\n",
    "      \"title\": \"Processing Level\"\n",
    "    },\n",
    "    \"publisher\": {\n",
    "      \"description\": \"Name of the organization or person responsible for making the file available.\",\n",
    "      \"title\": \"Publisher\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"sameAs\": {\n",
    "      \"description\": \"Another platform's URI or compact URI for this same resource. See https://www.w3.org/TR/owl-ref/#sameAs-def.\",\n",
    "      \"title\": \"Same As\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"sampleIdColumn\": {\n",
    "      \"description\": \"Name of the column containing sample identifiers\",\n",
    "      \"title\": \"Sample ID Column\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"source\": {\n",
    "      \"description\": \"Host platform from which the file was sourced.\",\n",
    "      \"title\": \"Source\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"species\": {\n",
    "      \"description\": \"Species of the organism(s) from which the data were generated. Omit property if not applicable, such as for data like compounds or other non-biological data.\",\n",
    "      \"items\": {\n",
    "        \"description\": \"\",\n",
    "        \"enum\": [\n",
    "          \"Rattus norvegicus\",\n",
    "          \"Gallus gallus\",\n",
    "          \"Pan troglodytes\",\n",
    "          \"Mus musculus (humanized)\",\n",
    "          \"Homo sapiens\",\n",
    "          \"Danio rerio\",\n",
    "          \"Drosophila melanogaster\",\n",
    "          \"Rhesus macaque\",\n",
    "          \"Sus scrofa\",\n",
    "          \"Oryctolagus cuniculus\",\n",
    "          \"Mus musculus\"\n",
    "        ],\n",
    "        \"title\": \"SpeciesEnum\",\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Species\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"subject\": {\n",
    "      \"description\": \"Applicable subject term(s) for file cataloging; use the Library of Congress Subject Headings (LCSH) scheme.\",\n",
    "      \"items\": {\n",
    "        \"type\": \"string\"\n",
    "      },\n",
    "      \"title\": \"Subject\",\n",
    "      \"type\": \"array\"\n",
    "    },\n",
    "    \"title\": {\n",
    "      \"title\": \"Title\",\n",
    "      \"type\": \"string\"\n",
    "    },\n",
    "    \"url\": {\n",
    "      \"description\": \"URL for the original file location at source.\",\n",
    "      \"title\": \"URL\",\n",
    "      \"type\": \"string\"\n",
    "    }\n",
    "  },\n",
    "  \"required\": [\n",
    "    \"title\",\n",
    "    \"creator\",\n",
    "    \"keywords\",\n",
    "    \"source\",\n",
    "    \"url\"\n",
    "  ]\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b9d4f85f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JSONSchema(name='OmicDataset', organization_name='ampals.schemas', organization_id='1415', id='9261', created_on='2026-02-10T05:25:09.007Z', created_by='3436666', uri='ampals.schemas-OmicDataset')\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/schema/type/create/async: 100%|██████████| 100/100 [00:01<00:00, 89.1it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'$schema': 'https://json-schema.org/draft-07/schema',\n",
       " '$id': 'https://repo-prod.prod.sagebase.org/repo/v1/schema/type/registered/ampals.schemas-OmicFile',\n",
       " 'type': 'object',\n",
       " 'properties': {'url': {'type': 'string',\n",
       "   'title': 'URL',\n",
       "   'description': 'URL for the original file location at source.'},\n",
       "  'assay': {'type': 'array',\n",
       "   'items': {'type': 'string',\n",
       "    'title': 'Assay',\n",
       "    'description': '',\n",
       "    'enum': ['2D AlamarBlue absorbance',\n",
       "     '2D AlamarBlue fluorescence',\n",
       "     '3D confocal imaging',\n",
       "     '3D electron microscopy',\n",
       "     '3D imaging',\n",
       "     '3D microtissue viability',\n",
       "     'actigraphy',\n",
       "     'AlgometRx Nociometer',\n",
       "     'auditory brainstem response',\n",
       "     'ATAC-seq',\n",
       "     'ATPase activity assay',\n",
       "     'BrdU proliferation assay',\n",
       "     'Bru-seq',\n",
       "     'BruChase-seq',\n",
       "     'CAPP-seq',\n",
       "     'CUT&RUN',\n",
       "     'ChIP-seq',\n",
       "     'Child Behavior Checklist for Ages 1.5-5',\n",
       "     'Child Behavior Checklist for Ages 6-18',\n",
       "     'CODEX',\n",
       "     'CROP-seq',\n",
       "     'confocal microscopy',\n",
       "     'Corsi blocks',\n",
       "     'current clamp assay',\n",
       "     'DART-seq',\n",
       "     'diffusion MRI',\n",
       "     'distortion product otoacoustic emissions',\n",
       "     'DNA optical mapping',\n",
       "     'ELISA',\n",
       "     'ERR bisulfite sequencing',\n",
       "     'EdU proliferation assay',\n",
       "     'FIA-MSMS',\n",
       "     'FLIPR high-throughput cellular screening',\n",
       "     'Fluorescence In Situ Hybridization',\n",
       "     'Focus group',\n",
       "     'FTIR spectroscopy',\n",
       "     'HI-C',\n",
       "     'HPLC',\n",
       "     'Interview',\n",
       "     'ISO-seq',\n",
       "     'MIB/MS',\n",
       "     'MeRIP-seq',\n",
       "     'Matrigel-based tumorigenesis assay',\n",
       "     'MudPIT',\n",
       "     'NIH Toolbox',\n",
       "     'NOMe-seq',\n",
       "     'RNA array',\n",
       "     'RNA-seq',\n",
       "     'RPPA',\n",
       "     'Riccardi and Ablon scales',\n",
       "     'SNP array',\n",
       "     'SUSHI',\n",
       "     'Sanger sequencing',\n",
       "     'Social Responsiveness Scale',\n",
       "     'Social Responsiveness Scale, Second Edition',\n",
       "     'SomaScan',\n",
       "     'T cell receptor repertoire sequencing',\n",
       "     'TIDE',\n",
       "     'TMT quantitation',\n",
       "     'TriKinetics activity monitoring',\n",
       "     'Von Frey test',\n",
       "     'active avoidance learning behavior assay',\n",
       "     'array',\n",
       "     'atomic force microscopy',\n",
       "     'autoradiography',\n",
       "     'bisulfite sequencing',\n",
       "     'blood chemistry measurement',\n",
       "     'blue native PAGE',\n",
       "     'body size trait measurement',\n",
       "     'bone histomorphometry',\n",
       "     'brightfield microscopy',\n",
       "     'cAMP-Glo Max Assay',\n",
       "     'CAGE-seq',\n",
       "     'calcium retention capacity assay',\n",
       "     'cell competition',\n",
       "     'cell count',\n",
       "     'cell painting',\n",
       "     'cell proliferation',\n",
       "     'cell viability assay',\n",
       "     'clinical data',\n",
       "     'CLIP-seq',\n",
       "     'cNF-Skindex',\n",
       "     'cognitive assessment',\n",
       "     'combination library screen',\n",
       "     'combination screen',\n",
       "     'complex II enzyme activity assay',\n",
       "     'compound screen',\n",
       "     'contextual conditioning behavior assay',\n",
       "     'conventional MRI',\n",
       "     \"Children's Dermatology Life Quality Index Questionnaire\",\n",
       "     'differential scanning calorimetry',\n",
       "     'dynamic light scattering',\n",
       "     'electrochemiluminescence',\n",
       "     'electrophoretic light scattering',\n",
       "     'elevated plus maze test',\n",
       "     'FACE-Q Appearance-related Distress',\n",
       "     'FACS',\n",
       "     'flow cytometry',\n",
       "     'focus forming assay',\n",
       "     'functional MRI',\n",
       "     'gait measurement',\n",
       "     'gel filtration chromatography',\n",
       "     'gel permeation chromatography',\n",
       "     'genotyping',\n",
       "     'high content screen',\n",
       "     'high frequency ultrasound',\n",
       "     'high-performance liquid chromatography/tandem mass spectrometry',\n",
       "     'immunoassay',\n",
       "     'immunocytochemistry',\n",
       "     'immunofluorescence',\n",
       "     'immunohistochemistry',\n",
       "     'in silico synthesis',\n",
       "     'in vitro tumorigenesis',\n",
       "     'in vivo PDX viability',\n",
       "     'in vivo bioluminescence',\n",
       "     'in vivo tumor growth',\n",
       "     'jumping library',\n",
       "     'LCM',\n",
       "     'label free mass spectrometry',\n",
       "     'laser speckle imaging',\n",
       "     'light scattering assay',\n",
       "     'liquid chromatography-electrochemical detection',\n",
       "     'liquid chromatography/mass spectrometry',\n",
       "     'liquid chromatography/tandem mass spectrometry',\n",
       "     'lncRNA-seq',\n",
       "     'local field potential recording',\n",
       "     'long term potentiation assay',\n",
       "     'mRNA counts',\n",
       "     'm6A array',\n",
       "     'magnetic resonance angiography',\n",
       "     'magnetic resonance spectroscopy',\n",
       "     'Magnetization-Prepared Rapid Gradient Echo MRI',\n",
       "     'mass spectrometry',\n",
       "     'massively parallel reporter assay',\n",
       "     'metabolic screening',\n",
       "     'methylation array',\n",
       "     'miRNA array',\n",
       "     'miRNA-seq',\n",
       "     'microrheology',\n",
       "     'Skindex-16',\n",
       "     'multi-electrode array',\n",
       "     'nanoparticle tracking analysis',\n",
       "     'NanoString nCounter Analysis System',\n",
       "     'n-back task',\n",
       "     'neuropsychological assessment',\n",
       "     'next generation targeted sequencing',\n",
       "     'novelty response behavior assay',\n",
       "     'open field test',\n",
       "     'optical tomography',\n",
       "     'optical coherence tomography',\n",
       "     'optokinetic reflex assay',\n",
       "     'oscillatory rheology',\n",
       "     'oxBS-seq',\n",
       "     'oxygen consumption assay',\n",
       "     'pattern electroretinogram',\n",
       "     'perineurial cell thickness',\n",
       "     'pharmocokinetic ADME assay',\n",
       "     'phase-contrast microscopy',\n",
       "     'photograph',\n",
       "     'polymerase chain reaction',\n",
       "     'polysomnography',\n",
       "     'positron emission tomography',\n",
       "     'PROMIS Cognitive Function',\n",
       "     'Olink Proximity Extension Assay',\n",
       "     'pure tone average',\n",
       "     'quantitative PCR',\n",
       "     'Quant-seq',\n",
       "     'questionnaire',\n",
       "     'reactive oxygen species assay',\n",
       "     'reporter gene assay',\n",
       "     'rheometry',\n",
       "     'RIP-seq',\n",
       "     'ribo-seq',\n",
       "     'rotarod performance test',\n",
       "     'sandwich ELISA',\n",
       "     'scCGI-seq',\n",
       "     'scale',\n",
       "     'SaferSeqS',\n",
       "     'single molecule drug screen assay',\n",
       "     'single-cell RNA-seq',\n",
       "     'single cell ATAC-seq',\n",
       "     'snATAC-seq',\n",
       "     'snRNA-seq',\n",
       "     'single-nucleus RNA-seq',\n",
       "     'small molecule library screen',\n",
       "     'sorbitol dehydrogenase activity level assay',\n",
       "     'spatial frequency domain imaging',\n",
       "     'spatial transcriptomics',\n",
       "     'static histomorphometry',\n",
       "     'static light scattering',\n",
       "     'survival',\n",
       "     'targeted exome sequencing',\n",
       "     'traction force microscopy',\n",
       "     'transcranial doppler ultrasonography',\n",
       "     'twin spot assay',\n",
       "     'ultra high-performance liquid chromatography/tandem mass spectrometry',\n",
       "     'western blot',\n",
       "     'whole exome sequencing',\n",
       "     'whole genome sequencing',\n",
       "     'whole-cell patch clamp',\n",
       "     'word recognition score',\n",
       "     'STR profile',\n",
       "     'Extension control 1',\n",
       "     'Incubation control 1',\n",
       "     'Amplification control 1',\n",
       "     'Protein target assay']},\n",
       "   'title': 'Measurement Technique',\n",
       "   'description': \"What's used to generate data in file, typically referring to the assay. Omit if not applicable.\"},\n",
       "  'title': {'type': 'string', 'title': 'Title'},\n",
       "  'sameAs': {'type': 'string',\n",
       "   'title': 'Same As',\n",
       "   'description': \"Another platform's URI or compact URI for this same resource. See https://www.w3.org/TR/owl-ref/#sameAs-def.\"},\n",
       "  'source': {'type': 'string',\n",
       "   'title': 'Source',\n",
       "   'description': 'Host platform from which the file was sourced.'},\n",
       "  'creator': {'type': 'array',\n",
       "   'items': {'type': 'string'},\n",
       "   'title': 'Creator',\n",
       "   'description': 'Properly formatted name of the organization or person that is creator of the file (e.g. \"Sage Bionetworks\" or \"John Doe\"), not an id. Most preferred is the project PI or data lead who\\'s responsible for its conception and initial content creation.'},\n",
       "  'disease': {'type': 'array',\n",
       "   'items': {'type': 'string',\n",
       "    'title': 'DiseaseEnum',\n",
       "    'description': '',\n",
       "    'enum': ['ALS', 'FTD', 'Control', 'Unknown']},\n",
       "   'title': 'Disease',\n",
       "   'description': 'Disease(s) or condition(s) studied in the file.'},\n",
       "  'license': {'title': 'License',\n",
       "   'description': 'Unless information for license is clear, this should default to UNKNOWN.'},\n",
       "  'species': {'type': 'array',\n",
       "   'items': {'type': 'string',\n",
       "    'title': 'SpeciesEnum',\n",
       "    'description': '',\n",
       "    'enum': ['Rattus norvegicus',\n",
       "     'Gallus gallus',\n",
       "     'Pan troglodytes',\n",
       "     'Mus musculus (humanized)',\n",
       "     'Homo sapiens',\n",
       "     'Danio rerio',\n",
       "     'Drosophila melanogaster',\n",
       "     'Rhesus macaque',\n",
       "     'Sus scrofa',\n",
       "     'Oryctolagus cuniculus',\n",
       "     'Mus musculus']},\n",
       "   'title': 'Species',\n",
       "   'description': 'Species of the organism(s) from which the data were generated. Omit property if not applicable, such as for data like compounds or other non-biological data.'},\n",
       "  'subject': {'type': 'array',\n",
       "   'items': {'type': 'string'},\n",
       "   'title': 'Subject',\n",
       "   'description': 'Applicable subject term(s) for file cataloging; use the Library of Congress Subject Headings (LCSH) scheme.'},\n",
       "  'cellType': {'type': 'array',\n",
       "   'items': {'type': 'string',\n",
       "    'title': 'CellTypeEnum',\n",
       "    'description': '',\n",
       "    'enum': ['motor_neuron',\n",
       "     'astrocyte',\n",
       "     'microglia',\n",
       "     'oligodendrocyte',\n",
       "     'fibroblast',\n",
       "     'pbmc',\n",
       "     'whole_blood',\n",
       "     'unknown']},\n",
       "   'title': 'Cell Type',\n",
       "   'description': 'Type of cell or cell line used in the file.'},\n",
       "  'citation': {'type': 'array',\n",
       "   'items': {'type': 'string'},\n",
       "   'title': 'Citation',\n",
       "   'description': 'identifies academic articles that are recommended by the data provider be cited in addition to the file itself.'},\n",
       "  'dataType': {'type': 'array',\n",
       "   'items': {'type': 'string',\n",
       "    'title': 'OmicDataTypeEnum',\n",
       "    'description': 'Types of omic data contained in files',\n",
       "    'enum': ['raw_sequencing',\n",
       "     'aligned_reads',\n",
       "     'variant_calls',\n",
       "     'genome_annotation',\n",
       "     'gene_expression',\n",
       "     'transcript_counts',\n",
       "     'differential_expression',\n",
       "     'splice_variants',\n",
       "     'methylation',\n",
       "     'chromatin_accessibility',\n",
       "     'histone_modification',\n",
       "     'protein_abundance',\n",
       "     'peptide_identification',\n",
       "     'protein_modification',\n",
       "     'metabolite_abundance',\n",
       "     'metabolic_pathways',\n",
       "     'integrated_omics',\n",
       "     'network_analysis',\n",
       "     'raw_data',\n",
       "     'processed_data',\n",
       "     'analyzed_results',\n",
       "     'proteomics',\n",
       "     'geneExpression',\n",
       "     'transcriptomics',\n",
       "     'transcriptomics_nascent',\n",
       "     'epigenomics',\n",
       "     'metabolomics',\n",
       "     'kinomics',\n",
       "     'isoformExpression',\n",
       "     'chromatinActivity',\n",
       "     'AlignedReads',\n",
       "     'genomicVariants',\n",
       "     'genomicFeatures',\n",
       "     'SomaticVariants',\n",
       "     'GermlineVariants',\n",
       "     'CopyNumberVariants',\n",
       "     'StructuralVariants',\n",
       "     'AnnotatedSomaticVariants',\n",
       "     'AnnotatedGermlineVariants',\n",
       "     'raw_counts',\n",
       "     'RawIntensities',\n",
       "     'NormalizedIntensities']},\n",
       "   'title': 'Data Type',\n",
       "   'description': 'Omic data types (genomics, transcriptomics, proteomics, etc.)'},\n",
       "  'keywords': {'type': 'array',\n",
       "   'items': {'type': 'string'},\n",
       "   'title': 'Keywords'},\n",
       "  'platform': {'type': 'string',\n",
       "   'title': 'Platform',\n",
       "   'description': 'Sequencing or analysis platform used',\n",
       "   'enum': ['7T Bruker Biospec',\n",
       "    '10x Visium Spatial Gene Expression',\n",
       "    '2D CellTiter-Glo',\n",
       "    '2D Incucyte',\n",
       "    'Affymetrix Genome-Wide Human SNP 5.0 Array',\n",
       "    'Affymetrix Genome-Wide Human SNP 6.0 Array',\n",
       "    'Affymetrix Human Gene 1.0 ST Array',\n",
       "    'Affymetrix Human Genome U133 Plus 2.0 Array',\n",
       "    'Affymetrix U133AB',\n",
       "    'Agilent 44Karray',\n",
       "    'BD FACS Calibur',\n",
       "    'BD FACSymphony',\n",
       "    'BGISEQ-500',\n",
       "    'Bionano Irys',\n",
       "    'Caliper',\n",
       "    'Cherry Imaging FACE Platform',\n",
       "    'Cherry Imaging TRACE Platform',\n",
       "    'Chromium X',\n",
       "    'EnVision 2103 Multiplate Reader',\n",
       "    'GE Discovery MR750 3T',\n",
       "    'GE Optima MR450W 1.5T',\n",
       "    'GE Signa HDxt 1.5T',\n",
       "    'GE Signa Genesis 1.5T',\n",
       "    'GE Signa HDxt 3T',\n",
       "    'GE Signa Premier 3T',\n",
       "    'GE Signa Excite 1.5T',\n",
       "    'Hitachi Echelon 1.5T',\n",
       "    'Hitachi Oasis 1.2T',\n",
       "    'IVIS Spectrum In Vivo Imaging System',\n",
       "    'Illumina 1M',\n",
       "    'Illumina Genome Analyzer IIx',\n",
       "    'Illumina HiSeq 2000',\n",
       "    'Illumina HiSeq 2500',\n",
       "    'Illumina HiSeq 3000',\n",
       "    'Illumina HiSeq 4000',\n",
       "    'Illumina HiSeq X',\n",
       "    'Illumina Human660W-Quad v1.0 BeadChip',\n",
       "    'Illumina HumanHap300',\n",
       "    'Illumina HumanMethylation450',\n",
       "    'Illumina HumanOmni1-Quadv1.0',\n",
       "    'Illumina HumanOmniExpress-24 v1.0 BeadChip',\n",
       "    'Illumina HumanOmniExpress-24 v1.2 BeadChip',\n",
       "    'Illumina Infinium MethylationEPIC BeadChip v1.0 (850k)',\n",
       "    'Illumina Infinium MethylationEPIC BeadChip v2.0 (935k)',\n",
       "    'Illumina MiSeq',\n",
       "    'Illumina MouseWG-6 v2.0 expression beadchip',\n",
       "    'Illumina NextSeq 1000',\n",
       "    'Illumina NextSeq 2000',\n",
       "    'Illumina NextSeq 500',\n",
       "    'Illumina NextSeq 550',\n",
       "    'Illumina NovaSeq 6000',\n",
       "    'Illumina NovaSeq X',\n",
       "    'Illumina NovaSeq X Plus',\n",
       "    'Illumina Omni2pt5M',\n",
       "    'Illumina Omni5M',\n",
       "    'Illumina WholeGenome DASL',\n",
       "    'Illumina h650',\n",
       "    'Infinium HumanOmniExpressExome',\n",
       "    'LI-COR Odyssey CLx',\n",
       "    'LTQ Orbitrap XL',\n",
       "    'Leica Aperio AT2',\n",
       "    'Leica MZ16',\n",
       "    'Leica S9 Stereomicroscope',\n",
       "    'LifeViz Infinity System',\n",
       "    'LifeViz Micro System',\n",
       "    'MGI T-series',\n",
       "    'Malvern Zetasizer',\n",
       "    'NanoFCM',\n",
       "    'NanoString Human nCounter PanCancer IO360 Panel',\n",
       "    'Nanostring Counter',\n",
       "    'Nanostring GeoMx',\n",
       "    'Not Applicable',\n",
       "    'Olympus DP80',\n",
       "    'Olympus IX73',\n",
       "    'Orbitrap Fusion Lumos Tribrid',\n",
       "    'Other Platform',\n",
       "    'Oxford Nanopore',\n",
       "    'PacBio RS II',\n",
       "    'PacBio Sequel II System',\n",
       "    'PacBio Sequel IIe System',\n",
       "    'Pannoramic 250 Flash',\n",
       "    'Perlegen 300Karray',\n",
       "    'Philips Achieva 1.5T',\n",
       "    'Philips Achieva 3T',\n",
       "    'Philips FEI Tecnai 12',\n",
       "    'Philips Intera Achieva 3T',\n",
       "    'Philips Ingenia 1.5T',\n",
       "    'Philips Ingenia 3T',\n",
       "    'Philips Panorama 1.0T',\n",
       "    'Promega GloMax Discover',\n",
       "    'Q Exative HF',\n",
       "    'Scale',\n",
       "    'Siemens Avanto 1.5T',\n",
       "    'Siemens Avanto Fit 1.5T',\n",
       "    'Siemens Magnetom Aera 1.5T',\n",
       "    'Siemens Magnetom Espree 1.5T',\n",
       "    'Siemens Magnetom Prisma 3T',\n",
       "    'Siemens Magnetom Skyra 3T',\n",
       "    'Siemens Magnetom Trio 3T',\n",
       "    'Siemens Magnetom Verio 3T',\n",
       "    'Siemens Magnetom Prisma Fit 3T',\n",
       "    'Spectramax M Series',\n",
       "    'TOOsonix System ONE-M',\n",
       "    'Toshiba Vantage Titan 1.5T',\n",
       "    'Varioskan LUX',\n",
       "    'Vectra H1 3D Imaging System',\n",
       "    'Ventana Benchmark XT',\n",
       "    'Vevo 3100 Imaging System',\n",
       "    'XF24 Extracellular Flux Analyzer',\n",
       "    'Zeiss LSM',\n",
       "    'Zeiss LSM 700',\n",
       "    'Zeiss LSM 980',\n",
       "    'Zeno Electronic Walkway',\n",
       "    'ZetaView']},\n",
       "  'publisher': {'type': 'string',\n",
       "   'title': 'Publisher',\n",
       "   'description': 'Name of the organization or person responsible for making the file available.'},\n",
       "  'collection': {'type': 'array',\n",
       "   'items': {'type': 'string',\n",
       "    'title': 'CollectionEnum',\n",
       "    'description': 'Original collection, e.g. the consortium that the dataset comes from.',\n",
       "    'enum': ['ALS Knowledge Portal',\n",
       "     'Answer ALS',\n",
       "     'CReATe Consortium',\n",
       "     'Genomic Translation for ALS Care (GTAC)',\n",
       "     'New York Genome Center ALS Consortium',\n",
       "     'National Institutes of Health',\n",
       "     'Project MinE',\n",
       "     'ALS Compute',\n",
       "     'amp_als',\n",
       "     'target_als',\n",
       "     'all_als',\n",
       "     'c_path',\n",
       "     'alsdb',\n",
       "     'geo',\n",
       "     'sra',\n",
       "     'als_compute']},\n",
       "   'title': 'Collection',\n",
       "   'description': 'Associated collection(s)'},\n",
       "  'fileFormat': {'type': 'string',\n",
       "   'title': 'File Format',\n",
       "   'description': 'Format of the omic data file (FASTQ, BAM, VCF, etc.)'},\n",
       "  'annotations': {'type': 'array',\n",
       "   'items': {'type': 'string',\n",
       "    'title': 'AnnotationTypeEnum',\n",
       "    'description': 'Types of biological databases and annotation references',\n",
       "    'enum': ['KEGG',\n",
       "     'GO',\n",
       "     'UniProt',\n",
       "     'Ensembl',\n",
       "     'HGNC',\n",
       "     'RefSeq',\n",
       "     'MONDO',\n",
       "     'HPO',\n",
       "     'SNOMED',\n",
       "     'ICD',\n",
       "     'pathway_enrichment',\n",
       "     'protein_pathway',\n",
       "     'metabolic_pathway',\n",
       "     'signaling_pathway']},\n",
       "   'title': 'Biological Annotations',\n",
       "   'description': 'Types of biological databases and annotations referenced in this file'},\n",
       "  'contributor': {'type': 'array',\n",
       "   'items': {'type': 'string'},\n",
       "   'title': 'Contributor(s)',\n",
       "   'description': 'Institution or person responsible for collecting and managing the files and records in the file.'},\n",
       "  'description': {'type': 'string',\n",
       "   'title': 'Description',\n",
       "   'description': 'Blurb for the file; should be no more than 500 characters.'},\n",
       "  'alternateName': {'type': 'string',\n",
       "   'title': 'Alternate Name',\n",
       "   'description': 'An alternate name that can be used for search and discovery improvement.'},\n",
       "  'analysisTypes': {'type': 'array',\n",
       "   'items': {'type': 'string',\n",
       "    'title': 'AnalysisMethodEnum',\n",
       "    'description': 'Computational and statistical analysis methods applied to the data',\n",
       "    'enum': ['Normalized Protein Expression (NPX)',\n",
       "     'GeneSetEnrichmentAnalysis',\n",
       "     'differential_expression',\n",
       "     'pathway_enrichment',\n",
       "     'network_analysis',\n",
       "     'proteomics_analysis',\n",
       "     'metabolomics_analysis',\n",
       "     'longitudinal_analysis',\n",
       "     'survival_analysis',\n",
       "     'meta_analysis',\n",
       "     'correlation_analysis',\n",
       "     'multivariate_analysis',\n",
       "     'machine_learning']},\n",
       "   'title': 'Analysis Types',\n",
       "   'description': 'Computational and statistical analysis methods applied to the data'},\n",
       "  'libraryLayout': {'type': 'array',\n",
       "   'items': {'type': 'string',\n",
       "    'title': 'LibraryLayoutEnum',\n",
       "    'description': '',\n",
       "    'enum': ['single', 'paired']},\n",
       "   'title': 'Library Layout',\n",
       "   'description': 'Single-end or paired-end sequencing layout for files containing sequencing data.'},\n",
       "  'FACSPopulation': {'type': 'string',\n",
       "   'title': 'FACS Population',\n",
       "   'description': 'FACS-sorted cell population description'},\n",
       "  'GEOSuperSeries': {'type': 'string',\n",
       "   'title': 'GEO SuperSeries',\n",
       "   'description': 'GEO SuperSeries identifier'},\n",
       "  'sampleIdColumn': {'type': 'string',\n",
       "   'title': 'Sample ID Column',\n",
       "   'description': 'Name of the column containing sample identifiers'},\n",
       "  'biospecimenType': {'type': 'array',\n",
       "   'items': {'type': 'string',\n",
       "    'title': 'BiospecimenTypeEnum',\n",
       "    'description': '',\n",
       "    'enum': ['blood',\n",
       "     'plasma',\n",
       "     'serum',\n",
       "     'csf',\n",
       "     'brain_tissue',\n",
       "     'spinal_cord',\n",
       "     'muscle',\n",
       "     'fibroblast',\n",
       "     'ipsc',\n",
       "     'motor_neuron',\n",
       "     'unknown']},\n",
       "   'title': 'Biospecimen Type',\n",
       "   'description': 'Type of biological specimen(s) in the file.'},\n",
       "  'individualCount': {'type': 'integer',\n",
       "   'title': 'Individual Count',\n",
       "   'description': 'Number of unique individuals included in the file (whether as individual-level or as aggregate data). Omit if not applicable/unknown.'},\n",
       "  'libraryStrategy': {'title': 'Library Strategy',\n",
       "   'description': 'Sequencing library strategy (RNA-seq, WGS, ChIP-seq, etc.)'},\n",
       "  'processingLevel': {'title': 'Processing Level',\n",
       "   'description': 'Level of data processing (raw, processed, analyzed)'},\n",
       "  'originalSampleName': {'type': 'string',\n",
       "   'title': 'Original Sample Name',\n",
       "   'description': 'Original sample name from data source'},\n",
       "  'includedInDataCatalog': {'type': 'string',\n",
       "   'title': 'Included In Data Catalog',\n",
       "   'description': 'Link(s) to known data catalog(s) the file is included in.'}},\n",
       " 'title': 'OmicFile',\n",
       " 'description': 'File containing omic data such as sequencing, expression, or proteomics data',\n",
       " 'required': ['title', 'creator', 'keywords', 'source', 'url'],\n",
       " 'additionalProperties': {}}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "organization = SchemaOrganization(name=ORG_NAME)\n",
    "try:\n",
    "    organization.store()\n",
    "except Exception as e:\n",
    "    organization.get()\n",
    "\n",
    "schemas = list(organization.get_json_schemas())\n",
    "for schema in schemas:\n",
    "    print(schema)\n",
    "\n",
    "schema = JSONSchema(name=SCHEMA_NAME, organization_name=ORG_NAME)\n",
    "try:\n",
    "    schema.get()\n",
    "except Exception as e:\n",
    "    schema.store(schema_body=schema_body, version=VERSION)\n",
    "\n",
    "schema.get_body()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bde6e2d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JSON schema was bound successfully. Please see details below:\n",
      "{'created_by': '3436666',\n",
      " 'created_on': '2026-02-10T05:32:11.885Z',\n",
      " 'id': 'ampals.schemas-OmicFile-0.0.1',\n",
      " 'json_sha256_hex': 'e2eda85d5bac2faab6ed71c002727307367719300e8c0731fc9784ccc9fe0a0e',\n",
      " 'organization_id': '1415',\n",
      " 'organization_name': 'ampals.schemas',\n",
      " 'schema_id': '9262',\n",
      " 'schema_name': 'OmicFile',\n",
      " 'semantic_version': '0.0.1',\n",
      " 'version_id': '52794'}\n"
     ]
    }
   ],
   "source": [
    "# Create a test folder for JSON schema experiments\n",
    "test_folder = Folder(name=\"test_folder\", parent_id=PROJECT_ENT.id).store()\n",
    "\n",
    "bound_schema = test_folder.bind_schema(\n",
    "    json_schema_uri=schema.uri, enable_derived_annotations=True\n",
    ")\n",
    "json_schema_version_info = bound_schema.json_schema_version_info\n",
    "syn.logger.info(\"JSON schema was bound successfully. Please see details below:\")\n",
    "pprint(vars(json_schema_version_info))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "14202a9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JSON Schema was retrieved successfully. Please see details below:\n",
      "{'created_by': '3436666',\n",
      " 'created_on': '2026-02-10T05:32:56.004Z',\n",
      " 'enable_derived_annotations': True,\n",
      " 'json_schema_version_info': JSONSchemaVersionInfo(organization_id='1415',\n",
      "                                                   organization_name='ampals.schemas',\n",
      "                                                   schema_id='9262',\n",
      "                                                   id='ampals.schemas-OmicFile-0.0.1',\n",
      "                                                   schema_name='OmicFile',\n",
      "                                                   version_id='52794',\n",
      "                                                   semantic_version='0.0.1',\n",
      "                                                   json_sha256_hex='e2eda85d5bac2faab6ed71c002727307367719300e8c0731fc9784ccc9fe0a0e',\n",
      "                                                   created_on='2026-02-10T05:32:11.885Z',\n",
      "                                                   created_by='3436666'),\n",
      " 'object_id': 73688145,\n",
      " 'object_type': 'entity'}\n"
     ]
    }
   ],
   "source": [
    "# 5. Retrieve the Bound Schema\n",
    "schema = test_folder.get_schema()\n",
    "syn.logger.info(\"JSON Schema was retrieved successfully. Please see details below:\")\n",
    "pprint(vars(schema))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "10fc8a82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validation statistics were retrieved successfully. Please see details below:\n"
     ]
    }
   ],
   "source": [
    "validation_statistics = test_folder.get_schema_validation_statistics()\n",
    "syn.logger.info(\n",
    "    \"Validation statistics were retrieved successfully. Please see details below:\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1b892a59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "See details of validation results: \n"
     ]
    }
   ],
   "source": [
    "invalid_validation = invalid_results = test_folder.get_invalid_validation()\n",
    "for child in invalid_validation:\n",
    "    syn.logger.info(\"See details of validation results: \")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "amp-als",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
